File: skvdb-client/src/main/java/org/skvdb/network/Connection.java
package org.skvdb.network;

import org.skvdb.common.service.UserService;
import org.skvdb.service.UserServiceImpl;
import org.skvdb.storage.Storage;

public class Connection {
    private String username;
    private String password;
    private NetworkService networkService;

    private Storage storage;
    private UserService userService;

    Connection(String host, int port, String username, String password) {
        networkService = new NetworkService(host, port);
        this.username = username;
        this.password = password;

        this.storage = new Storage(username, password, networkService);
        this.userService = new UserServiceImpl(username, password, networkService);
    }

    public Executor createExecutor() {
        return new Executor();
    }

    public void close() {
        networkService.close();
    }

    public class Executor {
        public Storage getStorage() {
            return storage;
        }

        public UserService getUserService() {
            return userService;
        }
    }
}

File: go-skvdb-client/src/conn/connection.go
package conn

import (
	"fmt"
	"go-skvdb-client/src/service/networkService"
	"go-skvdb-client/src/service/userService"
	"go-skvdb-client/src/storage"
)

type Connection struct {
	username       string
	password       string
	networkService *networkService.NetworkService

	storage     storage.Storage
	userService userService.UserService
}

func (conn *Connection) GetStorage() *storage.Storage {
	return &conn.storage
}

func (conn *Connection) GetUserService() *userService.UserService {
	return &conn.userService
}

func newConnection(host string, port int, username string, password string) (*Connection, error) {
	const op = "network.newConnection"
	ntwrkService, err := networkService.NewNetworkService(host, port)
	if err != nil {
		return nil, fmt.Errorf("%s : %w", op, err)
	}
	return &Connection{username: username, password: password, networkService: ntwrkService, storage: storage.NewStorage(ntwrkService, username, password), userService: userService.NewUserService(ntwrkService, username, password)}, nil
}
